### ISSUE TITLE
[DesktopGL] Suggestion: use FBO as back buffer render target

### ISSUE THREAD

**Issue by Jjagg** [MonoGame/MonoGame#6061]
Currently there are some issues in the DesktopGL implementation with respect to pixel format of the window. In OpenGL you can either use an FBO for MSAA and depth support or create the Default Framebuffer with support for MSAA and depth.

To create the Default FrameBuffer with MSAA and a depth/stencil buffer, you need to indicate this in the pixel format of the window that the context will be created for (e.g. on Windows by using [wglChoosePixelFormatArb](https://www.khronos.org/registry/OpenGL/extensions/ARB/WGL_ARB_pixel_format.txt)). Because of this you need to recreate the whole window whenever you want to change the surface format, depth bits, stencil bits or multisampling as there is no way to change the pixel format of an existing window [[1]](#setpixelformat). Because you recreate the window, you also need to recreate the GL context [[2]](#so-context). That in turn will mess up all graphics resources, and since recreating them isn't properly implemented yet this is a big issue. It's also annoying that you can actually see that the window is being recreated. 

If we use an FBO for MSAA and depth instead, the Default Framebuffer only needs a color buffer. We would render everything to the FBO instead of the Default Framebuffer and in `Present` blit the FBO color buffer to the Default Framebuffers color buffer. That way we can bypass the recreation of the window in most cases. The only case where we can't is if the backbuffer format is changed by the user because we need to change the color buffer of the Default Framebuffer [[3]](#bb-format). Using an FBO will allow changing multisample count, depth and stencil buffer size at runtime for DesktopGL. 
We'd still need to implement recreating the window in case the user changes the surface format, but using an FBO lowers priority for supporting window recreation IMO because changing the surface format at runtime is not something most users will do, while changing multisample count is more common.

Currently we create the Default Framebuffer with depth and MSAA support as required, however we do not support changing these settings at runtime because of the issues described in the second paragraph. 
So I suggest using an FBO instead. MG already requires FBO support for RenderTargets, but we would need require the [ARB_texture_multisample](https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_texture_multisample.txt) extension to have a multisampled FBO, so GL requirements if you want a multisampled back buffer would go up slightly [[4]](#requirements). I don't know how well supported ARB_texture_multisample is. The most significant benefit IMO is that we can let MG users change MSAA at runtime without any issues and without the player noticing (i.e. without recreating the window). We'd just need to switch out the FBO attachments, quick and simple.

We could optionally only use the FBO if ARB_texture_multisample is supported and fall back to creating the Default FrameBuffer with MSAA and depth buffer support, but that increases complexity and makes maintenance harder. (I think that's what FNA does, flibit calls the FBO 'faux-backbuffer' IIRC)

____________________

<a name="setpixelformat">[1]</a>: e.g. from [wgl docs](https://msdn.microsoft.com/en-us/library/dd369049(v=vs.85).aspx):
> Setting the pixel format of a window more than once can lead to significant complications for the Window Manager and for multithread applications, so it is not allowed. An application can only set the pixel format of a window one time. Once a window's pixel format is set, it cannot be changed.

<a name="so-context">[2]</a>: A GL context can apparently be used with the new window if the same pixel format is used, however that functionality is not exposed in SDL, see [this SO post](https://stackoverflow.com/a/12881231). And either way it's not really useful here, because we need to recreate the window *because* the pixel format needs to be changed.

<a name="bb-format">[3]</a>: Blitting does not require the formats of both involved color buffers to be the same, but it is limited in that not all combinations work and I think it's driver-dependent, so we should probably just ensure they're the same.

<a name="requirements">[4]</a>: If we decide to make RenderTargets optional to lower GL requirements (see #5580) this will make it a lot harder to do so.

______________________

**tl;dr**

Pros of FBO approach:

- Runtime switching of multisample count, depth and stencil buffer size without recreating the window.

Cons of FBO approach:

- Higher requirements if you want a multisampled back buffer (ARB_texture_multisample required)
- Harder to support older OpenGL versions if we ever decide we want that.

cc @cra0zy

**Comment by cra0zy** [comment]
So, the lesser of 2 evils.

**Comment by Jjagg** [comment]
I suppose :D

**Comment by KonajuGames** [comment]
To my mind, this is just having a hidden RenderTarget2D in GraphicsDevice
that is selected in the case of SetRenderTarget(null). During Present(),
this hidden RenderTarget2D is drawn as a quad to the back buffer. This
would also take care of colour format differences if needed. The window
colour format could be the same as the desktop, but the hidden revert
target could be different.

As for the pros and cons, I don't agree with going backwards in supporting
lower versions of OpenGL.

**Comment by mrhelmut** [comment]
ARB_texture_multisample is 3.1+, which is quite high to me (mostly because of Intel IGP missing a 3.0 driver on Win10). This is the only drawback that comes to my mind, otherwise this sounds interesting.

**Comment by Jjagg** [comment]
> To my mind, this is just having a hidden RenderTarget2D in GraphicsDevice that is selected in the case of SetRenderTarget(null). During Present(), this hidden RenderTarget2D is drawn as a quad to the back buffer.

Yes, that's pretty much how it is and maybe even how we'd implement it. Blitting with `glBlitFrameBuffer` is faster than drawing traditionally, but you're right that in case there's a format mismatch we could do a regular render instead so we don't have to recreate the window.

> ARB_texture_multisample is 3.1+, which is quite high to me

Yes, that is quite high. We could let support for changing ms count at runtime be optional and only allow setting it before window creation (i.e. in Game1 ctor) when ARB_texture_multisample is not supported. Could add a bool in GraphicsCapabilities for it. That way we don't sacrifice any functionality we have right now, but have better support with the extension and we don't need to implement (unless people really want it) the horrible mess that is window recreation.
